force-app/main/interoperability/classes/ProductRegistryVMapper.cls

```diff
@@ -0,0 +1,87 @@
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+ * @author Salesforce.org
+ * @date 2021
+ * @group
+ * @group-content
+ * @description Mapper class to transform ProductRegistryModel into ProductRegistryVModel
+ */
+public virtual with sharing class ProductRegistryVMapper {
+    @TestVisible
+    private static ProductRegistryVMapper instance;
+
+    /*********************************************************************
+     * @description Empty constructor is not public for Singleton Pattern
+     **********************************************************************/
+    protected ProductRegistryVMapper() {
+    }
+
+    /*********************************************
+     * @description Instance for Singleton Pattern
+     **********************************************/
+    public static ProductRegistryVMapper getInstance() {
+        if (instance == null) {
+            instance = new ProductRegistryVMapper();
+        }
+
+        return instance;
+    }
+
+    /********************************************************************************************
+     * @description Returns the list of ProductRegistryVModel based on the list of ProductRegistryModel
+     * @return   A list of ProductRegistryVModel with all its properties populated
+     ********************************************************************************************/
+    public virtual List<ProductRegistryVModel> getProductRegistrySettingsProductInformation() {
+        List<ProductRegistryVModel> productRegistryVModelList = new List<ProductRegistryVModel>();
+        List<ProductRegistryModel> productRegistryModelList = new List<ProductRegistryModel>();
+
+        productRegistryModelList = this.locateProductRegistryService().getProductRegistrySettingsProductInformation();
+
+        if (productRegistryModelList != null && !productRegistryModelList.isEmpty()) {
+            for (ProductRegistryModel prodRegistryModel : productRegistryModelList) {
+                ProductRegistryVModel prodRegistryVModel = new ProductRegistryVModel(
+                    prodRegistryModel.action,
+                    prodRegistryModel.namespace,
+                    prodRegistryModel.classname,
+                    prodRegistryModel.apiVersion
+                );
+                productRegistryVModelList.add(prodRegistryVModel);
+            }
+        }
```

Comment:
This is another place where we can make it abundantly clear what we expect from a unit testing perspective (and where another use case was hiding). Consider whether or not productRegistryModelList can ever be null, and if it can't, then we don't need to account for that use case. Though if it is, do we want to throw an exception?

If the list is empty, then the for loop won't do much harm from a processing perspective, since it will exit before it begins. So we don't need to list it in this case.

```suggestion
        if (productRegistryModelList == null) {
            return null;
        }

        for (ProductRegistryModel prodRegistryModel : productRegistryModelList) {
            ProductRegistryVModel prodRegistryVModel = new ProductRegistryVModel(
                prodRegistryModel.action,
                prodRegistryModel.namespace,
                prodRegistryModel.classname,
                prodRegistryModel.apiVersion
            );
            productRegistryVModelList.add(prodRegistryVModel);
        }
```